x/stakeibc/keeper/community_pool.go

```diff
@@ -1,12 +1,138 @@
+	hostZones := k.GetAllActiveHostZone(ctx)
+	for _, hostZone := range hostZones {
+		if hostZone.CommunityPoolDepositIcaAddress != "" &&
+			hostZone.CommunityPoolHoldingAddress != "" &&
+			hostZone.CommunityPoolReturnIcaAddress != "" {
+				// ICQ for the host denom of the chain, these are tokens the pool wants staked
+				err:= k.QueryCommunityPoolDepositBalance(ctx, hostZone, hostZone.HostDenom)
+				if err != nil {
+					k.Logger(ctx).Error(utils.LogWithHostZone(hostZone.ChainId, "Querying hostDenom %s - %s", hostZone.HostDenom, err.Error()))
+				}
+				// ICQ for the stToken of the host denom, these are tokens the pool wants redeemed
+				//   if stDenom is the denom on stride, ibcStDenom is the ibc denom on hostZone for stDenom
+				ibcStDenom := k.GetStakedHostTokenDenomOnHostZone(hostZone)
+				err = k.QueryCommunityPoolDepositBalance(ctx, hostZone, ibcStDenom)
+				if err != nil {
+					k.Logger(ctx).Error(utils.LogWithHostZone(hostZone.ChainId, "Querying stHostDenom %s - %s", ibcStDenom, err.Error()))
+				}				
+				// Transfer out all all tokens in the holding address back to the Return ICA
+				err = k.ReturnAllCommunityPoolTokens(ctx, hostZone)
+				if err != nil {
+					k.Logger(ctx).Error(utils.LogWithHostZone(hostZone.ChainId, "Returning from holding address - %s", err.Error()))
+				}				
+			}
+	}
+}
+
+
+// ICQ specific denom for balance in the deposit ICA on the community pool host zone
+// The ICQ callback will call IBCTransferCommunityPoolICATokensToStride with found token(s) as input
+func (k Keeper) QueryCommunityPoolDepositBalance(ctx sdk.Context, hostZone types.HostZone, denom string) error {
+	k.Logger(ctx).Info(utils.LogWithHostZone(hostZone.ChainId, "Submitting ICQ for %s in community pool deposit account balance", denom))
+
+	// Get the withdrawal account address from the host zone
+	if hostZone.CommunityPoolDepositIcaAddress == "" {
+		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no deposit account found for %s", hostZone.ChainId)
+	}
+
+	// Encode the deposit account address for the query request
+	// The query request consists of the account address and denom passed in
+	_, depositAddressBz, err := bech32.DecodeAndConvert(hostZone.CommunityPoolDepositIcaAddress)
+	if err != nil {
+		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "invalid withdrawal account address, could not decode (%s)", hostZone.CommunityPoolDepositIcaAddress)
+	}
+	queryData := append(bankTypes.CreateAccountBalancesPrefix(depositAddressBz), []byte(denom)...)
+
+	// The response might be a coin, or might just be an int depending on sdk version
+	// Since we need the denom later, store the denom as callback data for the query
+	callbackData := types.CommunityPoolDepositQueryCallback{
+		Denom: denom,
+	}
+	callbackDataBz, err := proto.Marshal(&callbackData)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to marshal community pool deposit balance callback data %v", callbackData)
+	}
+
+
+	// Timeout query at end of epoch
```

Comment:
I don't think there's any race condition here! If the packet times out, the channel closes anyways

Reply:
Still thinking about this -- I don't think we need to worry about the timeout packet racing with the ICA because the timeout policy is to reject query response so there would be no ICA triggered, but Sam and I have been discussing that the timeout for the the transfer which the ICA triggers on the other chain should be the epoch boundary for the 3 epoch plan to be easier to reason about.  

Therefore we should timeout this ICQ with at least enough time remaining in the epoch for the ICA and the transfer triggered by it to complete.  So maybe much earlier like 50% of the epoch or something for this ICQ timeout -- still should be plenty of time in most cases if we use the stride epochs

Reply:
My mental model is that either the relayer is on and both the ICQ and ICA will go through, or the relayer is off and neither will go through. 

That said, I think it makes the most sense to just time them both out at the end of the epoch. That way, if the relayer comes back on 3/4ths of the way through the epoch, they both get relayed (whereas with the 50/50 split, neither would)

Reply:
am i good to mark this resolved?